src/coreclr/jit/lower.cpp

```diff
@@ -5610,8 +5610,11 @@ void Lowering::InsertPInvokeMethodProlog()
     call->gtArgs.PushBack(comp, frameAddrArg);
 // for x86/arm32 don't pass the secretArg.
 #if !defined(TARGET_X86) && !defined(TARGET_ARM)
+    if (comp->lvaStubArgumentVar == BAD_VAR_NUM)
+        comp->lvaStubArgumentVar = comp->lvaInlinedPInvokeFrameVar;
```

Comment:
Without this `(comp->lvaStubArgumentVar == BAD_VAR_NUM)` cases lead to assert:
```gdb
Assert failure(PID 11606 [0x00002d56], Thread: 11606 [0x2d56]): Assertion failed 'lclNum < lvaCount' in 'System.Buffer:_Memmove(byref,byref,ulong)' during 'Lowering nodeinfo' (IL size 25; hash 0xd7d18401; FullOpts)

    File: /home/runtime/src/coreclr/jit/compiler.h:4008
    Image: /home/tmustafin/main_26mar_checkedO0_7f555e0_StubSecretArgRestored68d8aa/coreroot/corerun


Thread 1 "corerun" received signal SIGTRAP, Trace/breakpoint trap.
DBG_DebugBreak () at /home/runtime/src/coreclr/pal/src/arch/riscv64/debugbreak.S:6
6       /home/runtime/src/coreclr/pal/src/arch/riscv64/debugbreak.S: No such file or directory.
(gdb) bt
#0  DBG_DebugBreak () at /home/runtime/src/coreclr/pal/src/arch/riscv64/debugbreak.S:6
#1  0x0000003f74b129ae in DebugBreak () at /home/runtime/src/coreclr/pal/src/debug/debug.cpp:406
#2  0x0000003f7481c69a in assertAbort (why=0x3f74bc0f17 "lclNum < lvaCount", file=0x3f74bc17b0 "/home/runtime/src/coreclr/jit/compiler.h", line=4008) at /home/runtime/src/coreclr/jit/error.cpp:286
#3  0x0000003f7479cf10 in Compiler::lvaGetDesc (this=0x2aaac75278, lclNum=4294967295) at /home/runtime/src/coreclr/jit/compiler.h:4008
#4  0x0000003f7479cff2 in Compiler::lvaGetDesc (this=0x2aaac75278, lclVar=0x2aaac832d8) at /home/runtime/src/coreclr/jit/compiler.h:4020
#5  0x0000003f749e314c in Compiler::fgMorphLeafLocal (this=0x2aaac75278, lclNode=0x2aaac832d8) at /home/runtime/src/coreclr/jit/morph.cpp:4577
#6  0x0000003f749ef45a in Compiler::fgMorphLeaf (this=0x2aaac75278, tree=0x2aaac832d8) at /home/runtime/src/coreclr/jit/morph.cpp:8098
#7  0x0000003f749d9134 in Compiler::fgMorphTree (this=0x2aaac75278, tree=0x2aaac832d8, mac=0x0) at /home/runtime/src/coreclr/jit/morph.cpp:12721
#8  0x0000003f749da318 in Compiler::fgMorphArgs (this=0x2aaac75278, call=0x2aaac831b0) at /home/runtime/src/coreclr/jit/morph.cpp:3171
#9  0x0000003f749ebee0 in Compiler::fgMorphCall (this=0x2aaac75278, call=0x2aaac831b0) at /home/runtime/src/coreclr/jit/morph.cpp:7760
#10 0x0000003f749d9248 in Compiler::fgMorphTree (this=0x2aaac75278, tree=0x2aaac831b0, mac=0x0) at /home/runtime/src/coreclr/jit/morph.cpp:12746
#11 0x0000003f749e6052 in Compiler::fgMorphSmpOp (this=0x2aaac75278, tree=0x2aaac83398, mac=0x0, optAssertionPropDone=0x3fffff9f47) at /home/runtime/src/coreclr/jit/morph.cpp:8950
#12 0x0000003f749d9164 in Compiler::fgMorphTree (this=0x2aaac75278, tree=0x2aaac83398, mac=0x0) at /home/runtime/src/coreclr/jit/morph.cpp:12729
#13 0x0000003f749a51d8 in Lowering::InsertPInvokeMethodProlog (this=0x2aaac830c8) at /home/runtime/src/coreclr/jit/lower.cpp:5630
```

Reply:
Can you double check why `lvaStubArgumentVar` is `BAD_VAR_NUM` even if we are making use of this register? Does it mean we are adding garbage to this method call? It would be better to pass `0` as the argument in those cases.

Reply:
IIUC there are several different flavors of how the secret stub argument gets consumed:
- Through `System.StubHelpers.GetStubContext()` on all architectures. The runtime passes `CORJIT_FLAG_PUBLISH_SECRET_PARAM` for functions that may use this intrinsic.
- As an argument to `CORINFO_HELP_INIT_PINVOKE_FRAME` outside x86/arm32. The runtime seemingly doesn't always pass `CORJIT_FLAG_PUBLISH_SECRET_PARAM` in the cases where we need to call this JIT helper, which is confusing to me. Does it mean we can pass `nullptr` in these cases? Is the arg only needed for the pinvoke IL stubs, and not for normal functions with pinvokes in them?
- Through known stack frame layout on x86/arm32: https://github.com/dotnet/runtime/blob/b4765bd242a49cf932ae96b2bced86830de97b55/src/coreclr/vm/frames.h#L2850-L2880. Does anything stop us from unifying x86/arm32 with the rest of our targets and always passing it as an argument to the helper?

cc @jkotas, do you know the answer to some of these questions?

Reply:
> Is the arg only needed for the pinvoke IL stubs, and not for normal functions with pinvokes in them?

Yes, the arg is only used for IL stubs.

> Does it mean we can pass nullptr in these cases? 

The value is unused. It does not matter what gets passed in.

> Does anything stop us from unifying x86/arm32 with the rest of our targets and always passing it as an argument to the helper?

The JIT is aware of `InlinedFrame` layout and semantics. It may be better to switch 64-bit to the same scheme as x86/arm32: Get rid of the argument and store it in the `InlinedFrame` instead - on all platforms and only when it is needed. It will save a memory store for inlined PInvokes on x64.

Reply:
Thanks! I opened #100662 about unifying this.

@t-mustafin can you please update the fix to be something like
```cpp
    GenTree* argNode;
    if (comp->info.compPublishStubParam)
    {
        argNode = comp->gtNewLclvNode(comp->lvaStubArgumentVar, TYP_I_IMPL);
    }
    else
    {
        argNode = comp->gtNewIconNode(0, TYP_I_IMPL);
    }
    NewCallArg stubParamArg =
        NewCallArg::Primitive(argNode).WellKnown(WellKnownArg::SecretStubParam);
    call->gtArgs.PushBack(comp, stubParamArg);
```
?

Reply:
> I opened https://github.com/dotnet/runtime/issues/100662 about unifying this.

@t-mustafin In case you would like to fix #100662 and clean this up in this PR, here is the set of places:

- Delete code under the ifdef that you are fixing
- Delete StubSecretArg here: https://github.com/dotnet/runtime/blob/7def0b725852d55741262f43ed36ed501606f89f/src/coreclr/vm/jithelpers.cpp#L6069
- Delete https://github.com/dotnet/runtime/blob/7def0b725852d55741262f43ed36ed501606f89f/src/coreclr/vm/frames.h#L2893-L2898
- Enable this for all architectures https://github.com/dotnet/runtime/blob/7def0b725852d55741262f43ed36ed501606f89f/src/coreclr/vm/frames.h#L2856-L2871
- Rev the JIT-EE interface GUID: https://github.com/dotnet/runtime/blob/main/src/coreclr/inc/jiteeversionguid.h

@jakobbotsch Does this sound about right?

Reply:
I was hoping we would be able to get rid of the "known stack layout" contract between the VM and JIT as part of this unification. It results in a number of special casing around this particular value in the JIT and has a side effect that we cannot keep the value enregistered.
My thought was:
- Make `InlinedCallFrame::m_StubSecretArg` available on all platforms. Make it the JIT's responsibility to store this field in the frame on x64 (instead of doing it from `JIT_InitPInvokeFrame`). Store the value on x86 as well (it will change a spill to stack at the known stack frame location to a store to the `InlinedCallFrame`)
- Remove all special handling in frame layout around `lvaStubArgumentVar `
- Eventually enable full enregistration of `lvaStubArgumentVar` and remove all remaining special handling around it in the JIT. Coincidentally I am doing work in #100572 that should make this simple.

Do you see any potential issues around doing it this way?